sql-plugin/src/main/scala/com/nvidia/spark/rapids/CostBasedOptimizer.scala

```diff
@@ -270,6 +273,62 @@ class DefaultCostModel(conf: RapidsConf) extends CostModel {
 
 }
 
+/**
+ * Estimate the number of rows that an operator will output. Note that these row counts are
+ * the aggregate across all partitions.
+ *
+ * Logic is based on Spark's SizeInBytesOnlyStatsPlanVisitor. which operates on logical plans
+ * and only computes data sizes, not row counts.
+ */
+object RowCountPlanVisitor {
+
+  def visit(plan: SparkPlanMeta[_]): Option[BigInt] = plan.wrapped match {
+    case p: QueryStageExec =>
+      ShimLoader.getSparkShims.getQueryStageRuntimeStatistics(p).rowCount
+    case GlobalLimitExec(limit, _) =>
+      visit(plan.childPlans.head).map(_.min(limit)).orElse(Some(limit))
+    case LocalLimitExec(limit, _) =>
+      visit(plan.childPlans.head).map(_.min(limit)).orElse(Some(limit))
+    case p: SortMergeJoinExec =>
+      estimateJoin(plan, p.joinType)
+    case p: ShuffledHashJoinExec =>
+      estimateJoin(plan, p.joinType)
+    case p: BroadcastHashJoinExec =>
+      estimateJoin(plan, p.joinType)
+    case _: UnionExec =>
+      Some(plan.childPlans.flatMap(visit).sum)
+    case _ =>
```

Comment:
I would also want to add in something for aggregates.  Aggregates will change the number of output rows based off of the cardinality of the input, so it should be very similar to how we estimate a Join, with the key difference being that if there are no grouping keys (a reduction), then each partition will output a single row.

Reply:
I added in the empty grouping expression case and for other cases, it will fall through the default handling, which will just return the product of the child plan row counts. I don't know how we can get statistics on the cardinality of the grouping expressions to be able to determine anything more accurate. It would be nice if Spark could collect those kind of statistics in AQE.